### PR TITLE
fix duplicate primary keys for tables without models

### LINKED ISSUE #10354: 2.9 CakeSchema creates duplicate primary keys
* [x] bug
* CakePHP Version: latest 2.x (e.g. 2.9.6)
* Platform and Target: Platform and Target: Apache/2.2.31 (OSX/Homebrew), MySQL 5.7.17 (OSX/Homebrew) , PHP 7.0.15

### What you did

While using the `schema` shell to export my database schema and later on re-import it to my MySQL database I noticed I'm losing the AUTO_INCREMENT flag in a lot of my tables, which breaks the app.

### What happened

The reason for this is in `CakeSchema::_columns()`:
```php
	protected function _columns(&$Obj) {
		$db = $Obj->getDataSource();
		$fields = $Obj->schema(true);
		$columns = array();
		foreach ($fields as $name => $value) {
			if ($Obj->primaryKey === $name) {
				$value['key'] = 'primary';
			}
```
This above create a duplicate primary key for any table that has no model (which is supported by `CakeSchema` when you pass `models = false`) *and* a primary key already set on a *different* column than `id`.

When exporting a schema `CakeSchema::read()` will create an `AppModel` for each table, whether it is has an actual model or not:
```php
		$Object = new AppModel(array(
			'name' => Inflector::classify($table), 'table' => $table, 'ds' => $connection
		));
```
`read()` will then call `_columns()`. Since the `AppModel`'s `$primaryKey` is automatically set to `id` by the constructor at this point, the above code will trigger `$value['key'] = 'primary'`. However, since my table *already* has a primary key on a different column, re-importing the schema with a table that has now *two* primary keys seem to result with the `AUTO_INCREMENT` attribute being lost.

Here is a bit more context. I've *a lot* of these tables around, it would be painful to create a fake model for each one of them. Each of these tables is used to store *revisions* of tables that map to actual models, using the pretty handy [CakePHP-Assets/revision behavior by alkemann](https://github.com/alkemann/CakePHP-Assets/blob/master/models/behaviors/revision.php). For example `employees_revs` stores the revisions for the `Employee` model associated to the live `employees` table.

Revision models are created on the fly by the behavior as "shadow" models. The behavior also takes care of creating the revisions table which, per instructions, have the same fields as in the live table, except for: 
 * The `id` field should NOT be the primary key, nor auto increment
 * new fields `version_id` (int, primary key, autoincrement) and `version_created` (datetime) need to be present.

In other words, `id` is no longer primary or autoincrement because `version_id` is. This is important for this behavior to work, since we need `id` to match the `id` in the live table -- a record can (and will) have many revisions, therefore multiple rows will have the same `id` in the revisions table.

### What you expected to happen

I expected `id` not to be flagged as `primary` automatically.
I assume the behavior above was added for convenience?

Once fix would be to first check if the table already has *any* primary key. If that is the case, do not create a default primary key in the schema:
```php
	protected function _columns(&$Obj) {
		$db = $Obj->getDataSource();
		$fields = $Obj->schema(true);

		$hasPrimaryAlready = false;
		foreach ($fields as $value) {
			if (isset($value['key']) && $value['key'] === 'primary') {
				$hasPrimaryAlready = true;
			}
		}

		$columns = array();
		foreach ($fields as $name => $value) {
			if ($Obj->primaryKey === $name && !$hasPrimaryAlready && !isset($value['key'])) {
				$value['key'] = 'primary';
			}
```

Note that all tests pass with the above fix. I can create a PR but I've been scratching my head for an hour to create a test that exercises this bug specifically in `CakeSchemaTest.php`. I could use some help here, the test would need to call `read()` with `models => false` and a table need to be there already but without any model to match it against...

### PR DESCRIPTION
**Pull request by sebastienbarre**
This fixes #10354 for CakePHP 2.9

CakeSchema would create duplicate primary keys for tables that already a primary key set on a non-conventional column (e.g. not `id`).

### CODE REVIEW

**Comment by sebastienbarre** [comment]
Working on fixing coding standard and Postgres test, if I manage to install Postgres.

**Comment by sebastienbarre** [comment]
@markstory Unfortunately this seems unfixable for Postgres, so I had to skip the test for this datasource. Here is the current code in `CakeSchema`:
```php
	protected function _columns(&$Obj) {
		$db = $Obj->getDataSource();
		$fields = $Obj->schema(true);

		$hasPrimaryAlready = false;
		foreach ($fields as $value) {
			if (isset($value['key']) && $value['key'] === 'primary') {
				$hasPrimaryAlready = true;
			}
		}

		$columns = array();
		foreach ($fields as $name => $value) {
			if ($Obj->primaryKey === $name && !$hasPrimaryAlready && !isset($value['key'])) {
				$value['key'] = 'primary';
			}
```
Sadly, the call to `$Obj->schema(true);` is already borked when Postgres is in play. It returns a schema where:
* `id` is *already* hard-coded to be primary (so our new fix below with `$hasPrimaryAlready` doesn't matter),  
* `version_id` (the real primary) is not even marked as `primary` (as it is for the other backends), so that renders the fix even more moot.

The reason for this is this code in `Postgres.php`:
```php
public function describe($model) {
		$table = $this->fullTableName($model, false, false);
		$fields = parent::describe($table);
		$this->_sequenceMap[$table] = array();
		$cols = null;

		if ($fields === null) {
			$cols = $this->_execute(
				"SELECT DISTINCT table_schema AS schema, column_name AS name, data_type AS type, is_nullable AS null,
					column_default AS default, ordinal_position AS position, character_maximum_length AS char_length,
					character_octet_length AS oct_length FROM information_schema.columns
				WHERE table_name = ? AND table_schema = ?  ORDER BY position",
				array($table, $this->config['schema'])
			);

			// @codingStandardsIgnoreStart
			// Postgres columns don't match the coding standards.
			foreach ($cols as $c) {
[...]
				if ($model instanceof Model) {
					if ($c->name === $model->primaryKey) {
						$fields[$c->name]['key'] = 'primary';
```
There you have it, `if ($c->name === $model->primaryKey)` already forces `id` to be primary (for tables that have no explicitly implemented models). Now I try to see if I could loop over all the columns before that, just as I did in `CakeSchema`, to detect if there is a primary already, but no, that information is definitely not made available by `$cols`. To the extent that `version_id` is not even marked as `primary` here. 

Long story short, I don't think Postgres is supporting tables that have no models and use primary keys other than `id`. To my knowledge. This could probably be fixed with some clever joins in Postgres but I don't know anything about that backend.

**Comment by sebastienbarre** [comment]
@markstory is there a resident PostgreSQL expert for CakePHP who would want to fix that backend, or are we cool with this solution for 2.9.7? Works for all other backends. Thanks.

File: lib/Cake/Test/Case/Model/CakeSchemaTest.php [hunk 1]
```diff
@@ -650,6 +683,34 @@ public function testSchemaReadWithCrossDatabase() {
...
+		$read = $this->Schema->read(array(
+			'connection' => 'test',
+			'name' => 'TestApp',
+			'models' => false
+		));
+		$fixture->drop($db);
+
+		$hasTable = isset($read['tables']['non_conventional']);
+		$this->assertTrue($hasTable, 'non_conventional table should appear');
+		$versionIdHasKey = $hasTable && isset($read['tables']['non_conventional']['version_id']['key']);
```
Comment (markstory) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: You don't need this && here. If hasTable is not true the test will fail and no further assertions will be run.

**Comment by markstory** [comment]
I can take a look at why postgres is adding an id column. It might be a few days before I get to it though.

File: lib/Cake/Test/Case/Model/CakeSchemaTest.php (hunk 1 above)
Comment (sebastienbarre) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: Sure... as long as nothing happens after the assertions, or your test never fails. As it turns out I based my test on the one above, `testSchemaReadWithCrossDatabase()`. In this test, `$fixture->drop($db);` is placed *after* all the assertions. If any assertion fails, the fixture is not dropped and sticks around, and you won't be able to run the test anymore, you will get a `Error: Fixture creation for "non_conventional" failed "SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'non_conventional' already exists"` each time,  until you manually clean the database. That's why I added this `$hasTable` extra test... until I decided to move `$fixture->drop($db);` above all the assertions anyway -- so you are correct they are not needed anymore and I'll remove them. But the test above definitely has that problem...

Comment (markstory) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: I can take care of fixing the other test if you want.

Comment (sebastienbarre) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: Thanks

File: lib/Cake/Test/Case/Model/CakeSchemaTest.php [hunk 2]
```diff
@@ -364,6 +364,39 @@ class SchemaCrossDatabaseFixture extends CakeTestFixture {
 }
 
 /**
+ * NonConventionalPrimaryKeyFixture class
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: Nitpicking: No need for "class"

File: lib/Cake/Test/Case/Model/CakeSchemaTest.php [hunk 3]
```diff
@@ -364,6 +364,39 @@ class SchemaCrossDatabaseFixture extends CakeTestFixture {
...
+ *
+ * @package       Cake.Test.Case.Model
+ */
+class NonConventionalPrimaryKeyFixture extends CakeTestFixture {
+
+/**
+ * name property
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: Start with upper case. Same for the next two properties.

File: lib/Cake/Test/Case/Model/CakeSchemaTest.php (hunk 2 above)
Comment (ravage84) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: You _could_ add some short explanation what these fixtures are for. So one does not have to look up the related commit.

File: lib/Cake/Test/Case/Model/CakeSchemaTest.php (hunk 3 above)
Comment (sebastienbarre) [inline lib/Cake/Test/Case/Model/CakeSchemaTest.php]: what does?

**Comment by sebastienbarre** [comment]
Hey guys, just checking if this needs more work for 2.9.7 (due in one week?) or we are all good.

**Comment by markstory** [comment]
@sebastienbarre I still need to look into why postgres is always adding an id column.

File: lib/Cake/Model/CakeSchema.php [hunk 4]
```diff
@@ -613,9 +613,16 @@ protected function _columns(&$Obj) {
 		$db = $Obj->getDataSource();
 		$fields = $Obj->schema(true);
 
+		$hasPrimaryAlready = false;
+		foreach ($fields as $value) {
+			if (isset($value['key']) && $value['key'] === 'primary') {
+				$hasPrimaryAlready = true;
```
Comment (rellect) [inline lib/Cake/Model/CakeSchema.php]: I'd suggest breaking the loop once the primary is found..

Comment (sebastienbarre) [inline lib/Cake/Model/CakeSchema.php]: Done
